src/dune_pkg/fetch.ml

```diff
@@ -63,20 +63,28 @@ module Fiber_job = struct
     run1
 end
 
-let fetch loc url ~target =
+type failure =
+  | Checksum_mismatch of Checksum.t
+  | Unavailable of OpamUrl.t * string option
```

Comment:
The url here is always the input, right? If so, we don't need to return it. So what we're left with is an error message. We represent error messages with `User_message.t` so I would just use that.

Reply:
That might be a point, my main reasoning was that the `Not_found` exception is always very annoying because it doesn't specify what wasn't found. But exceptions are a little different than constructors.